### PR TITLE
Fixed #34981 -- Added pair and bisect options to DiscoverRunner.

### PR DESCRIPTION
**Pull request by sulabhkatila**
[ticket 34981](https://code.djangoproject.com/ticket/34981)

### CODE REVIEW

**Comment by nessita** [comment]
@sulabhkatila I'm not sure if you saw the [latest comment](https://code.djangoproject.com/ticket/34981#comment:10) in the ticket about this PR. Let us know if you have time to keep working on this! Thanks.

**Comment by sulabhkatila** [comment]
Hi @nessita I read the comment and will send send a revised PR soon

File: docs/releases/5.1.txt [hunk 1]
```diff
@@ -227,6 +227,13 @@ Tests
 * The Django test runner now supports a ``--screenshots`` option to save
   screenshots for Selenium tests.
 
+* The Django test runner now supports a ``--pair`` option that performs a pairwise
```
Comment (timgraham) [inline docs/releases/5.1.txt:230]: Please rephrase this as per https://github.com/django/django/pull/17799 (pending its merge).

File: docs/topics/testing/overview.txt [hunk 2]
```diff
@@ -385,3 +385,18 @@ Avoiding disk access for media files
...
+
+Detecting test pollution
+------------------------
+
+Executing tests using Bisection and Pairing
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes, a test may pass when run in isolation but fail when run as part of a
+test suite. This can be sue to test pollution, where some state from one test
+affects the execution of another test. To help analyze this problem, we have
```
Comment (timgraham) [inline docs/topics/testing/overview.txt:399]: "we have" -> ```"use the :option:`test --bisect` and :option:`test --pair` options.```

**Review by timgraham — COMMENTED** [review]
I didn't (and don't plan to) review this comprehensively, but here are a few notes.

File: django/test/utils.py [hunk 3]
```diff
@@ -381,6 +382,54 @@ def get_runner(settings, test_runner_class=None):
...
+            test_labels = test_labels_b[:-1]
+        elif failures_a.returncode and failures_b.returncode:
+            print("***** Multiple sources of failure found")
+            break
+        else:
+            print("***** No source of failure found... try pair execution (--pair)")
+            break
+
+    if len(test_labels) == 1:
+        print("***** Source of error: %s" % test_labels[0])
```
Comment (felixxm) [inline django/test/utils.py:430]: You copied a lot of code from `runtests.py`. As far as I'm aware, this patch should removed the matching code from `runtests.py` and use the functionality that we introduced in `DiscoverRunner`.

File: tests/runtests.py [hunk 4]
```diff
@@ -465,38 +470,11 @@ def bisect_tests(bisection_label, options, test_labels, start_at, start_after):
...
-        else:
-            print("***** No source of failure found... try pair execution (--pair)")
-            break
-
-    if len(test_labels) == 1:
-        print("***** Source of error: %s" % test_labels[0])
+    return run_bisection(
+        bisection_label,
+        test_labels,
+        subprocess_args,
```
Comment (felixxm) [inline tests/runtests.py:476]: We should be able to pass `pair` and `bisect` option to the `DiscoverRunner` and don't call `run_bisection()` and `run_pairing()` as an external tool.
